src/workerd/server/alarm-scheduler.c++

```diff
@@ -172,36 +172,36 @@ AlarmScheduler::ScheduledAlarm AlarmScheduler::scheduleAlarm(
 }
 
 kj::Promise<void> AlarmScheduler::checkTimestamp(kj::Duration delay, kj::Date scheduledTime) {
-  return timer.afterDelay(delay).then([this, scheduledTime]() mutable
-      -> kj::Promise<void> {
-    // Since we are waiting on timer.afterDelay, it's possible that timer.now() was behind
-    // the real time by a few ms, leading to premature alarm() execution. This checks it the current
-    // time is >= than scheduledTime to ensure we run alarms only on or after their scheduled time.
-    auto now = clock.now();
-    if (now >= scheduledTime) {
-      return kj::READY_NOW;
-    }
-    // If it's not yet time to trigger the alarm, we shall wait a while longer until we can trigger it.
-    // This repeats until it's time for the alarm to run.
-    kj::Duration additionalDelay = scheduledTime - now;
-    return checkTimestamp(additionalDelay, scheduledTime);
-  });
+  co_await timer.afterDelay(delay);
+
+  // Since we are waiting on timer.afterDelay, it's possible that timer.now() was behind
+  // the real time by a few ms, leading to premature alarm() execution. This checks it the current
+  // time is >= than scheduledTime to ensure we run alarms only on or after their scheduled time.
+  auto now = clock.now();
+  if (now < scheduledTime) {
+    // If it's not yet time to trigger the alarm, we shall wait a while longer until we can
+    // trigger it. This repeats until it's time for the alarm to run.
+    co_await checkTimestamp(scheduledTime - now, scheduledTime);
+  }
 }
 
-kj::Promise<void> AlarmScheduler::makeAlarmTask(
-    kj::Duration delay, const ActorKey& actorRef, kj::Date scheduledTime) {
-  return checkTimestamp(delay, scheduledTime)
-      .then([this, actor = actorRef.clone(), scheduledTime]() mutable -> kj::Promise<RetryInfo> {
-    {
-      auto& entry = KJ_ASSERT_NONNULL(alarms.findEntry(*actor));
-      entry.value.status = AlarmStatus::STARTED;
-    }
+kj::Promise<void> AlarmScheduler::makeAlarmTask(kj::Duration delay,
+                                                const ActorKey& actorRef,
+                                                kj::Date scheduledTime) {
+  co_await checkTimestamp(delay, scheduledTime);
 
-    return runAlarm(*actor, scheduledTime).catch_([](kj::Exception&& e)
-        -> kj::Promise<RetryInfo> {
-      KJ_LOG(WARNING, e);
+  {
+    auto& entry = KJ_ASSERT_NONNULL(alarms.findEntry(actorRef));
+    entry.value.status = AlarmStatus::STARTED;
+  }
 
-      return RetryInfo {
+  auto retryInfo = co_await ([&]() -> kj::Promise<RetryInfo> {
```

Comment:
It looks like this is safe because there's nothing in the catch handler that depends on the `&`-captured references, but the lack of `coCapture()`, or use of a captureless lambda`, makes me nervous.